### ISSUE TITLE
Android native bridge security concerns

### ISSUE THREAD

**Issue by EtienneLem** [apache/cordova-android#1064]
Quoting the [Iframes and the Callback Id Mechanism](https://cordova.apache.org/docs/en/latest/guide/appdev/security/index.html#iframes-and-the-callback-id-mechanism) section of the Security Guide

> If content is served in an iframe from a whitelisted domain, **that domain will have access to the native Cordova bridge**. This means that if you whitelist a third-party advertising network and serve those ads through an iframe, it is possible that a malicious ad will be able to break out of the iframe and perform malicious actions. Because of this, you should generally not use iframes unless you control the server that hosts the iframe content. Also note that there are third party plugins available to support advertising networks. Note that this statement is not true for iOS, which intercepts everything including iframe connections.

We are in a situation where we have iframes that can serve content from a third party source not controlled by us (integrations/plugins in our app). Our own app data is secured, the only way to communicate with our app is via [`window.postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). We even have 2 levels of iframe — it’s an iframe inside an iframe — where the first level is controlled by us (for CSP reasons).

When inspecting from the uncontrolled iframe, you can indeed have access to the [`_cordovaNative`](https://github.com/apache/cordova-android/blob/80b7a7f6acb43dfc2dfd4fd87261052496f37756/framework/src/org/apache/cordova/engine/SystemWebViewEngine.java#L235-L238) object which has the `exec` method. As you can see from the [method signature](https://github.com/apache/cordova-android/blob/master/cordova-js-src/android/promptbasednativeapi.js), it expects a `bridgeSecret` as the first parameter. Said secret seems to be generated from the native side and any attempt at using the method with the wrong secret will completely disable the bridge.

My question is: Since the secret isn’t exposed to JavaScript and it’s a different one for everyone and change every time you reload the app, how can this bridge exposure be exploited? I’d be also curious to know how plugins can still use `cordova.exec` without the bridge secret.

**Comment by breautek** [comment]
Javascript that executes in an iframe can access the parent's frame javascript environment via [window.frameElement](https://developer.mozilla.org/en-US/docs/Web/API/Window/frameElement)*.

> Since the secret isn’t exposed to JavaScript 

The secret is exposed/held in JavaScript memory. See https://github.com/apache/cordova-android/blob/565106fc1f840cb681905e6add8a44dce40dc984/cordova-js-src/exec.js#L112

Whether or not it's easily readable by code outside the cordova library is another question, but you should assume it's possible.

> I’d be also curious to know how plugins can still use cordova.exec without the bridge secret.

The secret is an abstraction and plugins don't need to be concerned about managing the bridge secret. They all call on an [platform-specific execute method](https://github.com/apache/cordova-android/blob/565106fc1f840cb681905e6add8a44dce40dc984/cordova-js-src/exec.js#L67) which does know the secret. So if untrusted code gains access to your cordova APIs, including plugin APIs, they'll be able to call on device APIs, so third-party code doesn't need to actually find the secret either to exploit your application.

Having remote/third-party code executing native APIs is also explicitly forbidden in [Apple's app store's terms of service](https://developer.apple.com/app-store/review/guidelines/#third-party-software), and [Google Play's app store's terms of service (Section Device and Network Abuse)](https://support.google.com/googleplay/android-developer/answer/9914283?hl=en). In this case, anything third-party I believe includes all executable code that isn't bundled within the application package itself.

*MDN says that the `frameElement` will be `null` if the iframe is cross-origin, however I'm not sure if it's the best idea to assume this is the case for all browsers or devices. I assume browsers that doesn't implement the CORS protocol doesn't do cross-origin checks, such as IE11 and older versions of the android webview that runs on android 4.4 - 5.0.

In conclusion, **assume your app is vulnerable** if you leak your cordova environment to third-party code.

Let me know if these answers your questions.

**Comment by EtienneLem** [comment]
Absolutely, that's pretty much how we interpreted the Security Guide. Thanks a lot @breautek

**Comment by breautek** [comment]
Can this issue be closed?

**Comment by EtienneLem** [comment]
Sure thing! Thanks again.

**Comment by vbraun** [comment]
If the cross-origin iframe can access `window.frameElement` then it is unsafe in the browser, too. Without cordova you may not be able to run native code but you can still steal authentication tokens, overlay malicious login screens, etc. So I don't understand why you shouldn't assume that `frameElement` is `null`, if it isn't then you have so many other problems before you even get to the cordova native bridge.
